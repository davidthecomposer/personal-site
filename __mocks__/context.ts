import { AppCtx } from "../src/types/context"

const mockAppCtx: AppCtx = true

export { mockAppCtx }
