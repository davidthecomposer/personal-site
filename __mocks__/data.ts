import after from "images/after.webp"

export const mockData = [
  {
    title: "Fake Title",
    image: after,
    slug: "",
    buttonText: "Hear it now",
    news: "Some thing about the news",
    audio: "",
    video: "",
  },
]
