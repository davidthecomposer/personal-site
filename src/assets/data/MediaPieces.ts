import flatline from "assets/images/flatline.jpg"
import after from "assets/images/after.jpg"
import forceOfNature from "assets/images/forceOfNature.jpg"
import mischeivous from "assets/images/mischeivous.jpg"
import redemption from "assets/images/redemption.jpg"
import rescue from "assets/images/rescue.jpg"
import starFlight from "assets/images/starFlight.jpg"
import starlight from "assets/images/starlight.jpg"
import turningPoint from "assets/images/turningPoint.jpg"
import loveStory from "assets/images/loveStory.jpg"
import flatlineM from "assets/images/flatlineM.jpg"
import afterM from "assets/images/afterM.jpg"
import forceOfNatureM from "assets/images/forceOfNatureM.jpg"
import mischeivousM from "assets/images/mischeivousM.jpg"
import redemptionM from "assets/images/redemptionM.jpg"
import rescueM from "assets/images/rescueM.jpg"
import starFlightM from "assets/images/starFlightM.jpg"
import starlightM from "assets/images/starlightM.jpg"
import turningPointM from "assets/images/turningPointM.jpg"
import loveStoryM from "assets/images/loveStoryM.jpg"

const flatlineMP3 = require("assets/audio/flatLine.mp3").default
const forceOfNatureMP3 = require("assets/audio/forceOfNature.mp3").default
const loveStoryMP3 = require("assets/audio/loveStory.mp3").default
const mischeivousMP3 = require("assets/audio/mischeivous.mp3").default
const redemptionMP3 = require("assets/audio/redemption.mp3").default
const rescueMP3 = require("assets/audio/rescue.mp3").default
const afterMP3 = require("assets/audio/rockIntro.mp3").default
const starlightMP3 = require("assets/audio/starlightStarbright.mp3").default
const starFlightMP3 = require("assets/audio/starflightStarbright.mp3").default
const turningPointMP3 = require("assets/audio/turningPoint.mp3").default

export const mediaPieces = [
  {
    title: "Flatline",
    video: false,
    audio: flatlineMP3,
    key: 0,
  },
  {
    title: "Rescue",
    video: false,
    audio: rescueMP3,
    key: 1,
  },
  {
    title: "Mischievous Endeavors",
    video: false,
    audio: mischeivousMP3,
    key: 2,
  },
  {
    title: "Turning Point",
    video: false,
    audio: turningPointMP3,
    key: 3,
  },
  {
    title: "After",
    video: false,
    audio: afterMP3,
    key: 4,
  },
  {
    title: "Redemption",
    video: false,
    audio: redemptionMP3,
    key: 5,
  },
  {
    title: "StarLight, StarFlight",
    img: [starlight, starlightM],
    video: false,
    audio: starlightMP3,
    key: 6,
  },
  {
    title: "Starflight, Starbright",
    img: [starFlight, starFlightM],
    video: false,
    music:
      "The choir was used to play off the immense sense of wonder and awe one might find when encountering something almost beyond understanding. I heavily processed the choir sounds during the last section of the piece to add to the sense of strangeness amidst the beauty of the main content.",
    story:
      "Our crew has been on a journey in hibernation for hundreds of years. As they wake up they are greeted by a view of dancing stars and twinkling lights as the engine progresses them toward their destination. As they approach the massive event horizon they are overpowered by a mix of awe, dread, and wonder. Small pieces of debris fly around them almost playfully as they slowly descend through to what lies beyond.",
    audio: starFlightMP3,
    key: 7,
  },
  {
    title: "Music for a Love Story",
    img: [loveStory, loveStoryM],
    video: false,
    music:
      "The music is unapologetically lyrical and simple. I tried to add as much color in the harmonies as I thought would be appropriate. The main little theme repeats over and over again through all instruments. I think that the soulful sound of the clarinet was a good choice as lead instrument here.",
    story:
      "A melancholy theme that touches on the kind of love that grows from long shared experience and sacrifice. A quiet montage of small kindnesses on a journey much nearer the end than the beginning.",
    audio: loveStoryMP3,
    key: 8,
  },
  {
    title: "Force Of Nature",
    img: [forceOfNature, forceOfNatureM],
    video: false,
    audio: forceOfNatureMP3,
    key: 9,
  },
]
