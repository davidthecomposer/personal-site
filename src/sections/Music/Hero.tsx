import React, { useEffect, useState, useContext, useMemo } from "react"
import styled from "styled-components"
import MainButton from "components/buttons/MainButton"
import text from "assets/styles/text"
import colors from "assets/styles/colors"
import gsap from "gsap"
import media from "assets/styles/media"
import { MobileContext } from "components/ContextStore"
import { IntroAnimationContext } from "pages/music"

import { navigate } from "gatsby"
import shadows from "assets/styles/shadows"
import { GatsbyImage } from "gatsby-plugin-image"

const Hero: React.FC<{ mobile: boolean; data: any }> = ({ data }) => {
  const [carouselWrapper, setCarouselWrapper] = useState<HTMLDivElement | null>(
    null
  )
  const [activeSlide, setActiveslide] = useState(0)
  const intro = useContext(IntroAnimationContext)
  const maxSlides = 5
  const mobile = useContext(MobileContext)
  const fakeNews = data.filter((news: any, i: number) => i < maxSlides - 1)

  useEffect(() => {
    if (carouselWrapper) {
      const tl = gsap.timeline({
        scrollTrigger: {
          trigger: carouselWrapper,
          toggleActions: "play pause resume none",
        },
        onComplete: () => setActiveslide(nextSlide),
      })
      const lastActive =
        activeSlide === 0 ? fakeNews.length - 1 : activeSlide - 1
      const nextSlide =
        activeSlide === fakeNews.length - 1 ? 0 : activeSlide + 1

      tl.to(
        `.slide_${activeSlide}`,
        {
          stagger: 0.35,
          opacity: 1,
          zIndex: 2,
          duration: 0.7,
          ease: "power2.inOut",
        },
        0.2
      )
        .to(
          `.slide_${lastActive}`,
          {
            stagger: 0.2,
            opacity: 0,
            duration: 0.3,
            zIndex: 1,
            ease: "power1.inOut",
          },
          0
        )
        .to(
          ".active-slide-num",
          {
            color: colors.activeTeal,
            opacity: 1,
            duration: 6.01,
            ease: "power1.inOut",
          },
          0
        )
        .to(".total-slide-num", { opacity: 1, duration: 0.5 }, 0)
        .to(
          ".active-slide-num",
          {
            color: colors.coolWhiteLight,
            opacity: 0,
            duration: 0.4,
            ease: "power1.inOut",
          },
          6.01
        )
        .to(
          ".slide_line",
          {
            scaleY: 1,
            scaleX: mobile ? 1 : 0,

            duration: 6,
          },
          0.1
        )
        .to(
          ".slide_line",
          {
            scaleY: mobile ? 1 : 0,
            scaleX: mobile ? 0 : 1,
            duration: 0.4,
          },
          6.2
        )

      return () => {
        tl.kill()
      }
    }
  }, [activeSlide, intro, carouselWrapper])

  const slides = useMemo(
    () =>
      fakeNews.map((news: any, i: number) => {
        console.log(news)
        return (
          <Slide key={`music_slide${i}`}>
            {news.audio && <audio src={news.audio} preload="none" />}
            {news.video && <video src={news.audio} preload="none" />}
            <Title className={`slide_${i}`}>{news.title}</Title>
            <MainImage
              className={`slide_${i}`}
              image={news.image[0].gatsbyImageData}
              alt={news.title}
            />
          </Slide>
        )
      }),
    [fakeNews]
  )

  const handleButton = (isNews: boolean, link: string) => {
    if (isNews) {
      navigate(`/news/${link}`)
    } else {
      window.open(link, "_blank")
    }
  }

  const buttons = fakeNews.map((news: any, i: number) => {
    return (
      <BtnWrap className={`slide_${i}`} key={`button_${i}`}>
        <MainButton
          onClick={() => handleButton(news.news, news.slug)}
          borderColor={colors.dullerTeal}
          backgroundColor={colors.activeTeal}
          limit
        >
          {news.buttonText}
        </MainButton>
      </BtnWrap>
    )
  })

  return (
    <Wrapper id="music-hero">
      <Carousel ref={ref => setCarouselWrapper(ref)}>
        <CarouselLine className="carousel_line1" top />
        <LeftPanel>
          <VertLine className="carousel_vert" />
          <LineTracker>
            <SlideNumber activeNumber={true} className="active-slide-num">
              0{activeSlide + 1}
            </SlideNumber>
            <Line className="slide_line" />
            <SlideNumber className="total-slide-num">
              0{fakeNews.length}
            </SlideNumber>
          </LineTracker>
          {buttons}
        </LeftPanel>
        {slides}
        <CarouselLine className="carousel_line2" />
      </Carousel>
    </Wrapper>
  )
}

const Wrapper = styled.section`
  padding: 12vw 0 0 0;
  position: relative;
  box-sizing: border-box;
  height: 62.5vw;
  ${media.fullWidth} {
    padding-top: 160px;
  }
  ${media.mobile} {
    width: 100%;
    overflow: hidden;
    padding: 15vw 0 58.1vw 0;
    height: 120.19vw;
  }
  ${media.tablet} {
    padding: 10vw 0 39.2vw 0;
  }
`

const CarouselLine = styled.div<{ top?: boolean | undefined }>`
  position: absolute;
  left: 0;
  width: 100%;
  transform: scaleX(0);
  height: 5px;
  background: ${colors.carouselTealDull};
  top: ${props => props.top && 0};
  bottom: ${props => !props.top && 0};
  transform-origin: ${props => (props.top ? "100% 100%" : "0% 0%")};

  ${media.mobile} {
    display: none;
  }
`

const Carousel = styled.div`
  width: 100%;
  height: 45.63vw;
  max-width: 1600px;
  max-height: 730px;
  position: relative;
  margin: 3.13vw auto 0;
  padding: 1.44vw 8.44vw 2.81vw 0;

  ${media.mobile} {
    width: 90%;
    height: 80vw;
    margin: 15vw auto 0;
  }
  ${media.fullWidth} {
    padding: 23.04px 135.04px 44.96px 0;
  }
`

const Slide = styled.div`
  width: 72.5vw;
  height: 41.38vw;
  position: absolute;
  top: 1.44vw;
  left: 18.94vw;

  ${media.mobile} {
    width: 100%;
    height: auto;
    left: 0;
    top: auto;
  }
  ${media.fullWidth} {
    width: 1160px;
    height: 662.08px;
    position: absolute;
    top: 23.04px;
    left: 303.04px;
  }
`

const Title = styled.h4`
  ${text.desktop.h4};
  color: ${colors.coolWhite};
  width: 100%;
  text-align: center;

  margin-bottom: 1vw;
  opacity: 0;

  ${media.mobile} {
    ${text.mobile.h4};
    margin-bottom: 4vw;
  }
  ${media.fullWidth} {
    ${text.fullWidth.h4};
    margin-bottom: 28px;
  }
`

const MainImage = styled(GatsbyImage)`
  width: 72.5vw;
  height: 36.63vw;
  border-radius: 1.205vw;
  opacity: 0;
  ${shadows.light};
  ${media.tablet} {
  }
  ${media.mobile} {
    width: 100%;
    height: auto;
  }
  ${media.fullWidth} {
    width: 1160px;
    height: 586px;
  }
`

const VertLine = styled.div`
  position: absolute;
  right: 0;
  top: 0;
  width: 4px;
  height: 100%;
  transform: scaleY(0);
  transform-origin: 0% 0%;
`

const LeftPanel = styled.div`
  width: 10.44vw;
  height: 40vw;
  position: absolute;
  left: 0;
  top: 2.81vw;

  ${media.mobile} {
    width: 100%;
    bottom: 0;
    top: auto;
    height: 16vw;
  }
  ${media.fullWidth} {
    width: 167.04px;
    height: 640px;
    top: 44.96px;
  }
`

const BtnWrap = styled.div`
  position: absolute;
  left: 1.31vw;
  bottom: 0;
  opacity: 0;
  z-index: 0;

  ${media.mobile} {
    top: 2vw;
    bottom: auto;
  }
`

const SlideNumber = styled.div<{ activeNumber?: boolean }>`
  ${text.desktop.bodyL};
  color: ${colors.coolWhiteLight};
  opacity: 0;

  ${media.mobile} {
    ${text.mobile.bodyL};
  }
  ${media.fullWidth} {
    ${text.fullWidth.bodyL};
  }
`

const LineTracker = styled.div`
  position: absolute;
  left: 3.75vw;
  top: 2.25vw;
  width: 2.75vw;
  height: 12vw;
  display: flex;
  flex-direction: column;
  align-items: center;

  ${media.mobile} {
    flex-direction: row-reverse;
    right: 4.67vw;
    left: auto;
    top: 0;
    width: 30.36vw;
    height: 100%;
  }
  ${media.fullWidth} {
    left: 60px;
    top: 36px;
    width: 44px;
    height: 192px;
  }
`

const Line = styled.div`
  width: 1px;
  height: 6.25vw;
  background: ${colors.carouselTeal};

  transform: scaleY(0);
  transform-origin: 100% 100%;

  ${media.mobile} {
    width: 20vw;
    height: 1px;
    transform: scaleY(1);
    transform: scaleX(0);
    transform-origin: 0% 0%;
    margin: 0 2vw;
  }
  ${media.fullWidth} {
    height: 100px;
  }
`

export default Hero
