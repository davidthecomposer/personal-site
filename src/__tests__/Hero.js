import React from "react"
import renderer from "react-test-renderer"
import { mockData } from "../../__mocks__/data"

import Hero from "sections/Music/Hero"

describe("Hero", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(<Hero data={mockData} mobile={false} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
