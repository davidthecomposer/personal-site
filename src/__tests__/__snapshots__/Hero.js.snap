// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<section
  className="sc-lknQiW cRFuly"
  id="music-hero"
>
  <div
    className="sc-dUipGc cwFbrU"
  >
    <div
      className="sc-jmpzUR bmVpoa carousel_line1"
    />
    <div
      className="sc-fWHiwC caVqzk"
    >
      <div
        className="sc-gUJyNl gKuDCh carousel_vert"
      />
      <div
        className="sc-gkSfol bXUKkD"
      >
        <div
          className="sc-gzrROc kZtABh active-slide-num"
        >
          0
          1
        </div>
        <div
          className="sc-fmPOXC bXclCH slide_line"
        />
        <div
          className="sc-gzrROc kZtABh total-slide-num"
        >
          0
          1
        </div>
      </div>
      <div
        className="sc-fIhvWL bCtTdS slide_0"
      >
        <button
          className="sc-gswNZR gusbqD"
          onClick={[Function]}
        >
          <span>
            Hear it now
          </span>
          <div
            className="sc-bcXHqe YMuFP"
          />
        </button>
      </div>
    </div>
    <div
      className="sc-hsWlPz knIrxI"
    >
      <h4
        className="sc-caPbAK bQlEgK slide_0"
      >
        Fake Title
      </h4>
      <img
        alt="Fake Title"
        className="sc-hjjlnc jOpAql slide_0"
        src="test-file-stub"
      />
    </div>
    <div
      className="sc-jmpzUR fCcOBm carousel_line2"
    />
  </div>
</section>
`;
